databricks/koalas/internal.py

```diff
@@ -657,96 +672,89 @@ def offset(id):
             (sdf[offset_column] + sdf[row_number_column] - 1).alias(column_name), *scols
         )
 
-    @lazy_property
-    def _column_labels_to_name(self) -> Dict[Tuple[str, ...], str]:
-        return dict(zip(self.column_labels, self.data_columns))
-
-    def column_name_for(self, column_labels_or_index_column: Union[str, Tuple[str, ...]]) -> str:
+    def spark_column_name_for(self, labels: Union[str, Tuple[str, ...]]) -> str:
         """ Return the actual Spark column name for the given column name or index. """
-        if column_labels_or_index_column in self._column_labels_to_name:
-            return self._column_labels_to_name[column_labels_or_index_column]
+        column_labels_to_name = dict(zip(self.column_labels, self.data_spark_column_names))
+        if labels in column_labels_to_name:
+            return column_labels_to_name[labels]  # type: ignore
         else:
-            if column_labels_or_index_column not in self.index_columns:
-                raise KeyError(name_like_string(column_labels_or_index_column))
-            return column_labels_or_index_column  # type: ignore
-
-    @lazy_property
-    def _column_labels_to_scol(self) -> Dict[Tuple[str, ...], spark.Column]:
-        return dict(zip(self.column_labels, self.column_scols))
+            for index_spark_column_name, index_name in self.index_map.items():
+                if index_name == labels:
+                    return index_spark_column_name
+            raise KeyError(name_like_string(labels))
 
-    def scol_for(self, column_labels_or_index_column: Union[str, Tuple[str, ...]]):
+    def spark_column_for(self, labels: Union[str, Tuple[str, ...]]):
         """ Return Spark Column for the given column name or index. """
-        if column_labels_or_index_column in self._column_labels_to_scol:
-            return self._column_labels_to_scol[column_labels_or_index_column]
+        column_labels_to_scol = dict(zip(self.column_labels, self.data_spark_columns))
+        if labels in column_labels_to_scol:
+            return column_labels_to_scol[labels]  # type: ignore
         else:
-            if column_labels_or_index_column not in self.index_columns:
-                raise KeyError(name_like_string(column_labels_or_index_column))
-            return scol_for(self._sdf, self.column_name_for(column_labels_or_index_column))
+            if labels not in self.index_spark_column_names:
+                raise KeyError(name_like_string(labels))
+            return scol_for(self._sdf, self.spark_column_name_for(labels))
```

Comment:
We can't get index spark column properly.

Is this `label` for index, string as index column name or tuple as index name?
If string, it will fail when `self.spark_column_name_for(labels)`, otherwise it will fail when `labels not in self.index_spark_column_names`.

Reply:
I was thinking `spark_column_for` should only search based on labels or index name, not the actual Spark name. I thought the previous version mixes both external (pandas-like labels) and internal (Spark column name). For example, for regular series, it requires labels to search instead of the physical column names.

The current version, it searches only by pandas-like labels or pandas-like index name now. For Spark columns, it has to be manually searched against Spark column names via:

```diff
- kdf._internal.scol_for(index_spark_column_name)
+ scol_for(kdf.spark_frame, index_spark_column_name)
```

For pandas-like search via labels or index name, it has to always be:

```python
kdf._internal.scol_for(labels_or_index_name)
```

Does this make sense to you?

Reply:
The fact that the current change doesn't cause test failures makes me wonder whether we need to keep support index columns in these three methods. I mean, those three methods are only for looking for data columns by the column labels.

Reply:
The following change doesn't cause test failure. We don't use `spark_column_name_for`, `spark_column_for`, and `spark_type_for` for looking for index columns at all with the current changes.

```diff
$ git diff
diff --git a/databricks/koalas/internal.py b/databricks/koalas/internal.py
index 6c09033..b9a1590 100644
--- a/databricks/koalas/internal.py
+++ b/databricks/koalas/internal.py
@@ -674,14 +674,7 @@ class _InternalFrame(object):

     def spark_column_name_for(self, labels: Union[str, Tuple[str, ...]]) -> str:
         """ Return the actual Spark column name for the given column name or index. """
-        column_labels_to_name = dict(zip(self.column_labels, self.data_spark_column_names))
-        if labels in column_labels_to_name:
-            return column_labels_to_name[labels]  # type: ignore
-        else:
-            for index_spark_column_name, index_name in self.index_map.items():
-                if index_name == labels:
-                    return index_spark_column_name
-            raise KeyError(name_like_string(labels))
+        return self._sdf.select(self.spark_column_for(labels)).columns[0]

     def spark_column_for(self, labels: Union[str, Tuple[str, ...]]):
         """ Return Spark Column for the given column name or index. """
@@ -689,9 +682,7 @@ class _InternalFrame(object):
         if labels in column_labels_to_scol:
             return column_labels_to_scol[labels]  # type: ignore
         else:
-            if labels not in self.index_spark_column_names:
-                raise KeyError(name_like_string(labels))
-            return scol_for(self._sdf, self.spark_column_name_for(labels))
+            raise KeyError(name_like_string(labels))

     def spark_type_for(self, labels: Union[str, Tuple[str, ...]]) -> DataType:
         """ Return DataType for the given column name or index. """
```

How about we drop support for looking for index columns with these methods?
Maybe the way to look for index columns is different from for data columns, then we should have another method for it?

Reply:
+100!